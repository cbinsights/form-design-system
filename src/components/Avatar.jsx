import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../util/baseElement';

const VALID_LIGHT_COLORS = ['white', 'haze', 'lightGray'];
const VALID_DARK_COLORS = ['orange', 'charcoal', 'navy', 'aqua', 'gray', 'red', 'purple'];
export const VALID_BG_COLORS = [...VALID_LIGHT_COLORS, ...VALID_DARK_COLORS];
export const SIZES = ['s', 'm'];

export const grabInitials = (str) =>
  str
    .split(' ')
    .slice(0, 2)
    .reduce((prev, curr) => prev + curr.charAt(0), '')
    .toUpperCase();

const Avatar = ({ bgColor, imgUrl, size, name, Link, ...rest }) => {
  const Element = baseElement({ href: rest.href, onClick: rest.onClick, as: Link });
  return (
    <Element
      {...rest}
      role="img"
      title={name}
      className={cx('fdsAvatar', `bgColor--${bgColor}`, `fdsAvatar--${size}`, {
        'border--focus': Element !== 'div',
        'color--white': VALID_DARK_COLORS.includes(bgColor),
      })}
    >
      {imgUrl && (
        <span className="fdsAvatar-img" style={{ backgroundImage: `url(${imgUrl})` }} />
      )}
      {name && grabInitials(name)}
    </Element>
  );
};

Avatar.defaultProps = {
  bgColor: 'purple',
  size: 'm',
};

Avatar.propTypes = {
  /** Controls color of button */
  bgColor: PropTypes.oneOf(VALID_BG_COLORS),
  /** Controls the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Control initials displayed, and also used for title accessibility attribute (uses 2 words max for initials) */
  name: PropTypes.string.isRequired,
  /** Sets background image over initials */
  imgUrl: PropTypes.string,
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
};

export default Avatar;
