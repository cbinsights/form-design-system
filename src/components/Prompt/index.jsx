import React from 'react';
import PropTypes from 'prop-types';
import Dialog from 'components/Dialog';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';

const Prompt = (props) => (
  <Dialog
    content={<div id="a11y-dialog-desc">{props.desc}</div>}
    footerContent={
      <Flex justify="end" wrap>
        {props.secondaryButton && <FlexItem shrink>{props.secondaryButton}</FlexItem>}
        {props.primaryButton && <FlexItem shrink>{props.primaryButton}</FlexItem>}
      </Flex>
    }
    role="alertdialog"
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
  />
);

Prompt.propTypes = {
  /** Controls text in header */
  title: PropTypes.string,
  /** Controls string displayed in content area */
  desc: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  /** Represents primary button (MUST be an FDS Button) */
  primaryButton: PropTypes.any.isRequired,
  /** Represents secondary button (MUST be an FDS Button) */
  secondaryButton: PropTypes.any.isRequired,
  /** Controls whether modal (and backdrop) are visible or not */
  isOpen: PropTypes.bool,
  /**
   * Callback that user can pass in, to be conditionally fired when
   * user attempts to close modal. When defined, the modal close button appears
   */
  onDismiss: PropTypes.func,
};

export default Prompt;
