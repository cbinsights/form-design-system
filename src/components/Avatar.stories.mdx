import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, text, optionsKnob as options } from '@storybook/addon-knobs';
import { arrayToOptions } from './util/storybook';

import Avatar, { VALID_BG_COLORS, SIZES } from './Avatar';

export const bgColorOptions = options('bgColor', arrayToOptions(VALID_BG_COLORS), undefined, {
  display: 'inline-radio',
});

<Meta title="Components|Avatar" decorators={[withKnobs]} component={Avatar} />

# Avatar

<Story name="Knobs">
  <Avatar
    bgColor={options('bgColor', arrayToOptions(VALID_BG_COLORS), undefined, {
      display: 'inline-radio',
    })}
    size={options('size', arrayToOptions(SIZES), undefined, {
      display: 'inline-radio',
    })}
    name={text('name', 'Samuel Jackson')}
    imgUrl={text('imgUrl', undefined)}
  />
</Story>

<Props of={Avatar} />

### Colors
Accepted colors are based off FDS `bgColor--{color}` classes: https://cbinsights.github.io/form-design-system/fds-styles/#background-colors

<Preview>
  <Story name="Colors">
    <div className="display--inlineFlex margin--all" style={{ flexWrap: 'wrap' }}>
      <div className="margin--right--half margin--bottom--half">
        <Avatar name="default" />
      </div>
      {VALID_BG_COLORS.map(bgColor => (
        <div className="margin--right--half margin--bottom--half">
          <Avatar name={`${bgColor.slice(0,1)} ${bgColor.slice(1,2)}`} bgColor={bgColor} />
        </div>
      ))}
      <br />
    </div>
  </Story>
</Preview>

### Images
Initials will show either if you don't pass an image, the image doesn't load, or it hasn't loaded yet.

<Preview>
  <Story name="Variations">
    <div className="display--inlineFlex margin--all" style={{ flexWrap: 'wrap' }}>
      <div className="margin--right--half margin--bottom--half">
        <Avatar name="Hello Impecable" imgUrl="http://placeimg.com/80/81/animals" />
      </div>
      <div className="margin--right--half margin--bottom--half">
        <Avatar name="Hello Impecable" imgUrl="invalid url" />
      </div>
      <br />
    </div>
  </Story>
</Preview>

### Sizes

<Preview>
  <Story name="Sizes">
    <div className="display--inlineFlex margin--all" style={{ flexWrap: 'wrap' }}>
      {SIZES.map(size => (
        <div className="margin--right--half margin--bottom--half">
          <Avatar name="Hello Impecable " size={size} />
        </div>
      ))}
      <br />
    </div>
  </Story>
</Preview>