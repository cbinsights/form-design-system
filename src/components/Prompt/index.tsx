import React from 'react';
import Dialog from 'components/Dialog';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import { ButtonProps } from 'components/Button';

export interface PromptProps {
  /** Controls text in header */
  title?: string;
  /** Controls string displayed in content area */
  desc?: React.ReactNode | string;
  /** Represents primary button (MUST be an FDS Button) */
  primaryButton: React.ReactElement<ButtonProps>;
  /** Represents secondary button (MUST be an FDS Button) */
  secondaryButton: React.ReactElement<ButtonProps>;
  /** Controls whether modal (and backdrop) are visible or not */
  isOpen?: boolean;
  /**
   * Callback that user can pass in, to be conditionally fired when
   * user attempts to close modal. When defined, the modal close button appears
   */
  onDismiss?: () => void;
  /** Keeps the Prompt from being rendered in a portal and renders it locally instead. */
  disablePortal?: boolean;
}

const Prompt = (props: PromptProps): JSX.Element => (
  <Dialog
    content={props.desc ? <div id="a11y-dialog-desc">{props.desc}</div> : null}
    footerContent={
      <Flex justify="end">
        {props.secondaryButton && <FlexItem shrink>{props.secondaryButton}</FlexItem>}
        {props.primaryButton && <FlexItem shrink>{props.primaryButton}</FlexItem>}
      </Flex>
    }
    role="alertdialog"
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
    disablePortal={props.disablePortal}
  />
);

export default Prompt;
