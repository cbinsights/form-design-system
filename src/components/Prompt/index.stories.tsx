import React from 'react';
import { action } from '@storybook/addon-actions';
import { NoStoriesLayout } from 'util/storybook-docs/Layout';
import Button from 'components/Button';

import Prompt from '.';

const Template = (args) => <Prompt {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  isOpen: true,
  desc: 'This change will be permanent and cannot be undone.',
  title: 'Are you sure you want to delete this collection?',
  primaryButton: <Button isDestructive label="Delete" />,
  secondaryButton: (
    <Button theme="ghost" label="Keep Collection" onClick={action('close modal')} />
  ),
  onDismiss: action('close modal'),
};

export const NoDescription = Template.bind({});
NoDescription.args = {
  ...Primary.args,
  desc: false,
};

export default {
  component: Prompt,
  title: 'components/Prompt',
  parameters: {
    componentSubtitle: 'An inflexible version of Dialog that is used for confirm modals.',
    docs: {
      fds: {
        related: ['Dialog'],
      },
      page: NoStoriesLayout,
      description: {
        component:
          'Displays a simple string message, with 2 buttons to either confirm the action, or deny it. **Please click "Canvas" to view examples**.',
      },
    },
  },
};
