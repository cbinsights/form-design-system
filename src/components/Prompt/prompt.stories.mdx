import {
  Meta,
  Story,
  Preview,
  Props,
  Description
} from "@storybook/addon-docs/blocks";
import { withKnobs, text, boolean, object } from "@storybook/addon-knobs";
import { action } from '@storybook/addon-actions';
import { ImportPath } from 'components/util/storybook';
import Button from "components/Button";

import Prompt from "./";

export const primaryButtonProps = () =>
  object("primaryButton", {
    isDestructive: true,
    label: "Yes, delete"
  });

export const secondaryButtonProps = () =>
  object("secondaryButton", {
    theme: "ghost",
    label: "Keep Collection",
    onClick: action("close modal"),
    }
  );

<Meta
  title="Components/Prompt"
  decorators={[withKnobs]}
  component={Prompt}
  parameters={{ docs: { disable: true } }}
/>

# Prompt

<ImportPath component={Prompt} section="modals" />

*Please click **"Canvas"** to view examples.*

A confirm modal that displays a simple message, with 2 buttons to either confirm the action, or deny it.

<Props of={Prompt} />

<Story name="Knobs">
  <Prompt
    isOpen={boolean("isOpen", true)}
    desc={text(
      "content",
      "This change will be permanent and cannot be undone."
    )}
    title={text("title", "Are you sure you want to delete this collection?")}
    primaryButton={<Button {...primaryButtonProps()} />}
    secondaryButton={<Button {...secondaryButtonProps()} />}
    onDismiss={action('close modal')}
  />
</Story>
