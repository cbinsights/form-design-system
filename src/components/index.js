import Avatar from 'components/Avatar';
import Badge from 'components/Badge';
import Button from 'components/Button';
import ButtonGroup from 'components/ButtonGroup';
import Checkbox from 'components/Checkbox';
import Chip from 'components/Chip';
import DateInput from 'components/DateInput';
import Dialog from 'components/Dialog';
import DropdownButton from 'components/DropdownButton';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import FloatingAction from 'components/FloatingAction';
import Hscroll from 'components/Hscroll';
import IconButton from 'components/IconButton';
import InputGroup from 'components/InputGroup';
import Menu from 'components/Menu';
import MenuItem from 'components/MenuItem';
import Popover from 'components/Popover';
import Prompt from 'components/Prompt';
import Radio from 'components/Radio';
import StackedButton from 'components/StackedButton';
import TextInput from 'components/TextInput';
import Toaster from 'components/Toaster';
import Tooltip from 'components/Tooltip';
import Updatable from 'components/Updatable';

export {
  Avatar,
  Badge,
  Button,
  ButtonGroup,
  Checkbox,
  Chip,
  DateInput,
  Dialog,
  DropdownButton,
  Flex,
  FlexItem,
  FloatingAction,
  Hscroll,
  IconButton,
  InputGroup,
  Menu,
  MenuItem,
  Popover,
  Prompt,
  Radio,
  StackedButton,
  TextInput,
  Toaster,
  Tooltip,
  Updatable,
};
