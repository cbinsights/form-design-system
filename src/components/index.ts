import Avatar from 'components/Avatar';
import Badge from 'components/Badge';
import Button from 'components/Button';
import ButtonGroup from 'components/ButtonGroup';
import Checkbox from 'components/Checkbox';
import Chip from 'components/Chip';
import CornerPosition from 'components/CornerPosition';
import DateInput from 'components/DateInput';
import Dialog from 'components/Dialog';
import DropdownButton from 'components/DropdownButton';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import FloatingAction from 'components/FloatingAction';
import IconButton from 'components/IconButton';
import Indicator from 'components/Indicator';
import InputGroup from 'components/InputGroup';
import Menu from 'components/Menu';
import MenuItem from 'components/MenuItem';
import MenuLink from 'components/MenuLink';
import Popover from 'components/Popover';
import Prompt from 'components/Prompt';
import Radio from 'components/Radio';
import Search from 'components/TopSearch';
import SeparatorList from 'components/SeparatorList';
import StackedButton from 'components/StackedButton';
import TextInput from 'components/TextInput';
import Toaster from 'components/Toaster';
import Tooltip from 'components/Tooltip';
import ZeroState from 'components/ZeroState';

export {
  Avatar,
  Badge,
  Button,
  ButtonGroup,
  Checkbox,
  Chip,
  CornerPosition,
  DateInput,
  Dialog,
  DropdownButton,
  Flex,
  FlexItem,
  FloatingAction,
  IconButton,
  Indicator,
  InputGroup,
  Menu,
  MenuItem,
  MenuLink,
  Popover,
  Prompt,
  Radio,
  Search,
  SeparatorList,
  StackedButton,
  TextInput,
  Toaster,
  Tooltip,
  ZeroState,
};
