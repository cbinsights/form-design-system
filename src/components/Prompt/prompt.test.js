import React from 'react';
import { shallow } from 'enzyme';
import { Button } from 'components';
import Prompt from '.';

describe('Prompt component', () => {

  it('matches snapshot (default props)', () => {
    const component = shallow(<Prompt primaryButton={<Button label="foo" />} secondaryButton={<Button label="bar" />} />);
    expect(component).toMatchSnapshot();
    const component2 = shallow(<Prompt isOpen primaryButton={<Button label="foo" />} secondaryButton={<Button label="bar" />} />);
    expect(component2).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = shallow(
      <Prompt 
        primaryButton={<Button label="foo" />} 
        secondaryButton={<Button label="bar" />} 
        title="title"
        desc="desc"
        isOpen
      />
    );
    expect(component).toMatchSnapshot();
  });

})