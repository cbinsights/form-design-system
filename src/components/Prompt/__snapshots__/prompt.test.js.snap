// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt component matches snapshot (default props) 1`] = `
<Dialog
  content={
    <div
      id="a11y-dialog-desc"
    />
  }
  footerContent={
    <Flex
      justify="end"
      wrap={true}
    >
      <FlexItem
        shrink={true}
      >
        <Button
          label="bar"
        />
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <Button
          label="foo"
        />
      </FlexItem>
    </Flex>
  }
  role="alertdialog"
/>
`;

exports[`Prompt component matches snapshot (default props) 2`] = `
<Dialog
  content={
    <div
      id="a11y-dialog-desc"
    />
  }
  footerContent={
    <Flex
      justify="end"
      wrap={true}
    >
      <FlexItem
        shrink={true}
      >
        <Button
          label="bar"
        />
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <Button
          label="foo"
        />
      </FlexItem>
    </Flex>
  }
  isOpen={true}
  role="alertdialog"
/>
`;

exports[`Prompt component matches snapshot (set all props) 1`] = `
<Dialog
  content={
    <div
      id="a11y-dialog-desc"
    >
      desc
    </div>
  }
  footerContent={
    <Flex
      justify="end"
      wrap={true}
    >
      <FlexItem
        shrink={true}
      >
        <Button
          label="bar"
        />
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <Button
          label="foo"
        />
      </FlexItem>
    </Flex>
  }
  isOpen={true}
  role="alertdialog"
  title="title"
/>
`;
